zio-http/src/main/scala/zio/http/Request.scala

```diff
@@ -1,199 +1,54 @@
+   * Drops trailing slash from the path.
    */
-  def body: Body
-
-  def copy(
-    version: Version = self.version,
-    method: Method = self.method,
-    url: URL = self.url,
-    headers: Headers = self.headers,
-  ): Request = {
-    val m = method
-    val u = url
-    val h = headers
-    val v = version
-    new Request {
-      override final val method: Method   = m
-      override final val url: URL         = u
-      override final val headers: Headers = h
-      override final val version: Version = v
-      override final val body: Body       = self.body
+  def dropTrailingSlash: Request = self.copy(url = self.url.dropTrailingSlash)
 
-      override final val unsafe: UnsafeAPI = new UnsafeAPI {
-        override final def context(implicit unsafe: Unsafe): ChannelHandlerContext = self.unsafe.context
-        override final def encode(implicit unsafe: Unsafe): HttpRequest            = self.unsafe.encode
-      }
-    }
-  }
+  val path = url.path
 
-  /**
-   * Remove trailing slash from path.
-   */
-  final def dropTrailingSlash: Request = self.copy(url = self.url.dropTrailingSlash)
+  def withMethod(method: Method): Request = self.copy(method = method)
 
-  /**
-   * Gets all the headers in the Request
-   */
-  def headers: Headers
+  def withPath(path: Path): Request = self.copy(url = self.url.copy(path = path))
 
-  /**
-   * Checks is the request is a pre-flight request or not
-   */
-  def isPreflight: Boolean = method == Method.OPTIONS
-
-  /**
-   * Gets the request's method
-   */
-  def method: Method
-
-  /**
-   * Gets the request's path
-   */
-  def path: Path = url.path
-
-  /**
-   * Overwrites the method in the request
-   */
-  final def setMethod(method: Method): Request = self.copy(method = method)
+  def withUrl(url: URL): Request = self.copy(url = url)
 
   /**
-   * Overwrites the path in the request
-   */
-  final def setPath(path: Path): Request = self.copy(url = self.url.copy(path = path))
-
-  /**
-   * Overwrites the url in the request
-   */
-  final def setUrl(url: URL): Request = self.copy(url = url)
-
-  /**
-   * Gets the complete url
-   */
-  def url: URL
-
-  /**
-   * Gets the request's http protocol version
+   * Updates the headers using the provided function
    */
-  def version: Version
-
-  private[zio] trait UnsafeAPI {
-
-    /**
-     * Accesses the channel's context for more low level control
-     */
-    def context(implicit unsafe: Unsafe): ChannelHandlerContext
-
-    /**
-     * Gets the HttpRequest
-     */
-    def encode(implicit unsafe: Unsafe): HttpRequest
-  }
-
-  private[zio] val unsafe: UnsafeAPI
+  override def updateHeaders(update: Headers => Headers): Request = self.copy(headers = update(self.headers))
 }
 
 object Request {
 
   /**
-   * Constructor for Request
+   * Constructor for Request that provides sane defaults.
    */
-  def apply(
+  def make(
     version: Version = Version.`HTTP/1.1`,
     method: Method = Method.GET,
     url: URL = URL.root,
```

Comment:
Is `URL.root` really a sane default?

Reply:
Let me be more specific. I think a URL can/should not have any default value. I think the other values are sane, because they have some technical reason(ing) behind them. And they describe the _how_. But URL is the _what_ and I'd assume, that in 99.9% of the cases URL won't be `URL.root` and will be overwritten.
Idk if this is the right PR to do it (most likely not), but since we are working on a 2.0 release with a lot of changes anyway, I'd say we should change this.
I would say, maybe one could just remove the default URL parameter, or have methods to build a request based on the http method somewhat like this
```scala
object Request {
...
  def get(url:URL): Request = make(url = url)
  def post(url: URL, body: Body) = make(url = url, body = body)
```
And then the idiomatic use would be
```scala
val myRequest = Request.get(url).addHeaders(myHeaders)
```
@jdegoes or @vigoo maybe you have an opinion? I am happy to open an issue/pr, if you see some potential for improvement 🙂

Reply:
@jdegoes Rebased and pushed.